Improve getting started (#87)
juliushaertl/nextcloud-docker-dev: README.md

./bootstrap.sh
sudo sh -c "echo '127.0.0.1 nextcloud.local' >> /etc/hosts"
docker compose up nextcloud proxy
```

Ok, let's go to understand these commands line !

First, you download the project with the git command, then you move to the `nextcloud-docker-dev` folder.

The `bootstrap.sh` script check if all requirements are present and prepares your workspace.
You have the `./workspace` folder where there is the `server` folder, the Nextcloud's core and [other Nextcloud versions](docs/running-stable-versions.md) if you want (the stable21, stable22, stable23, and so on).

So, you if you want to contribute to the Nextcloud's core, you can work in this folder directly !

Then, you add `nextcloud.local` to your hosts file.

Finally, the `docker compose up nextcloud proxy` command line. This command line runs the nextcloud, proxy, redis and mailhog containers.

## First connection

After running the `docker compose up nextcloud proxy` command, you have to wait a few minutes before trying to connect to your development instance.

In fact, you will see that the `nextcloud` and `proxy` containers initialize your Nextcloud, create  user accounts, and so on. Step by step.

> Then, once these steps are completed, you can connect to your development instance. You must enter `http://nextcloud.local` only in your search bar !

Reviewer: ```suggestion
Then, once these steps are completed, you can connect to your development instance. You must enter `http://nextcloud.local` in your address bar!
```

```

You can generate selfsigned certificates using:
### Which user accounts can I use ?

```
cd data/ssl
openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout  nextcloud.local.key -out nextcloud.local.crt
```
Here is a list of user accounts you can use :

### dnsmasq to resolve wildcard domains
| uid | password |
|:---:|:---:|
| admin | admin |
| user1 | user1 |
| user2 | user2 |
| user3 | user3 |
| user4 | user4 |
| user5 | user5 |
| user6 | user6 |
| nextcloud | nextcloud |
| alice | alice |
| bob | bob |
| jane | jane |
| john | john |

Instead of adding the individual container domains to `/etc/hosts` a local dns server like dnsmasq can be used to resolve any domain ending with the configured DOMAIN_SUFFIX in `.env` to localhost.
## Where can I add my app for development ?

For dnsmasq adding the following configuration would be sufficient for `DOMAIN_SUFFIX=.local`:
```
address=/.local/127.0.0.1
```

### Use valid certificates trusted by your system

* Install mkcert https://github.com/FiloSottile/mkcert
* Go to `data/ssl`
* `mkcert nextcloud.local`

* `mv nextcloud.local-key.pem nextcloud.local.key`
* `mv nextcloud.local.pem nextcloud.local.crt`

> * `docker-compose restart proxy`

Reviewer: All SSL related info seems gone, whether selfsigned or this version?
Browsers may very well soon drop HTTP support, it’s important to show how to get HTTPS support.

Author: No, it didn't disappear.

I move this section to `docs/containers/ssl.md`. But, I forgot to add a link here : https://github.com/arawa/nextcloud-docker-dev/blob/13a001965c2e7472474295aa361fdd26baa0a204/README.md?plain=1#L89 ^^'

I am going to fix it ^^'

Author: I fixed it here : https://github.com/juliushaertl/nextcloud-docker-dev/commit/b501b22a3b9d9b23baf20c0d523d460bac789d44

It's okay for you, or do you feel that something needs to be changed ? Don't hesitate to tell me :)

	AuthType Basic
	AuthName "SAML"
	AuthUserFile /var/www/html/.htpasswd
	Require valid-user
</Location>
```

## Fulltextsearch

```
docker-compose up -d elasticsearch elasticsearch-ui
```

- Address for configuring in Nextcloud: `http://elastic:elastic@elasticsearch:9200`
- Adress to access elastic search from outside: `http://elastic:elastic@elasticsearch.local`
- Address for accessing the ui: http://elasticsearch-ui.local/

`sudo sysctl -w vm.max_map_count=262144`

## Object storage

Primary object storage can be enabled by setting the `PRIMARY=minio` environment variable either in your .env file or in docker-compose.yml for individual containers.

```bash
docker-composer up proxy nextcloud minio
```

## Development

### OCC

Run inside of the Nextcloud container:
```
set XDEBUG_CONFIG=idekey=PHPSTORM
sudo -E -u www-data php -dxdebug.remote_host=192.168.21.1 occ
```

### Useful commands

- Restart apache to reload php configuration without a full container restart: `docker-compose kill -s USR1 nextcloud`
- Access to mysql console: `mysql -h $(docker inspect -f '{{range.NetworkSettings.Networks}}{{.IPAddress}}{{end}}' nextcloud_database-mysql_1) -P 3306 -u nextcloud -pnextcloud`

> - Run an LDAP search: `ldapsearch -x -H ldap://$(docker inspect -f '{{range.NetworkSettings.Networks}}{{.IPAddress}}{{end}}' nextcloud_ldap_1) -D "cn=admin,dc=planetexpress,dc=com" -w admin -b "dc=planetexpress,dc=com" -s subtree <filter> <attrs>`

Reviewer: Where are these gone?

Author: I moved some of this documentation to the `setup-xdebug.md` file.

I added a link in the README.md file : https://github.com/arawa/nextcloud-docker-dev/blob/13a001965c2e7472474295aa361fdd26baa0a204/README.md?plain=1#L78

Author: do you feel that something needs to be changed ? Don't hesitate to tell me :)

# nextcloud-dev-docker-compose

Nextcloud development environment using docker-compose
Hi beginner developer ! 👋

> ⚠ **DO NOT USE THIS IN PRODUCTION** Various settings in this setup are considered insecure and default passwords and secrets are used all over the place

Reviewer 2: I'd love to keep this note as people should never expose this environment to the public or run it anywhere except their local and hopefully secured machines.

Author: Okay, I rewrite your text in the README.md :)

I specify that this project is for the development environment only here : https://github.com/arawa/nextcloud-docker-dev/blob/66251329279419ebaf5d69147240c1e55dec24d5/README.md?plain=1#L8

Maybe is insufficient ? :thinking:

Author: I fixed your request in this commit : https://github.com/juliushaertl/nextcloud-docker-dev/pull/87/commits/7442538c2f81bd270dbb58acec2c691b3c654918

> This project allows you to get started a nextcloud server and develop your first app !

Reviewer 2: ```suggestion
This project allows you to start developing a Nextcloud app or contribute to Nextcloud server.
```

Author: I could not accept your commit suggestion from UI :/

I apply your request in this commit : https://github.com/juliushaertl/nextcloud-docker-dev/pull/87/commits/201da9717829a8645c02f02619e542d54ca0b154

So, you can use this project for your development environment only.

If you don't know how to install Docker, please, read this documentation : https://cloud.nextcloud.com/s/iyNGp8ryWxc7Efa .
Be careful : Read the "README.md" first.

## Getting started 🎮

First, get the setup running:

```bash
git clone https://github.com/juliushaertl/nextcloud-docker-dev
cd nextcloud-docker-dev
./bootstrap.sh
sudo sh -c "echo '127.0.0.1 nextcloud.local' >> /etc/hosts"
docker compose up nextcloud proxy
```

Ok, let's go to understand these commands line !

First, you download the project with the git command, then you move to the `nextcloud-docker-dev` folder.

The `bootstrap.sh` script check if all requirements are present and prepares your workspace.
You have the `./workspace` folder where there is the `server` folder, the Nextcloud's core and [other Nextcloud versions](docs/running-stable-versions.md) if you want (the stable21, stable22, stable23, and so on).

So, you if you want to contribute to the Nextcloud's core, you can work in this folder directly !

Then, you add `nextcloud.local` to your hosts file.

> Finally, the `docker compose up nextcloud proxy` command line. This command line runs the nextcloud, proxy, redis and mailhog containers.

Reviewer 2: Should we also give an example how to stop with Ctrl+C and that you can run it in the background with `docker compose up -d ...`?

Author: Yes, I am writing how to stop with Ctrl+C. But, I don't know to run the containers in the background with the `-d` flag.

Because, it was my first error to run your project. Because, you see all information to understand what happen and you can connect on the instance when you see all users created :

```bash
nextcloud-nextcloud-1  |    The user "alice" was created successfully
nextcloud-nextcloud-1  |    The user "user6" was created successfully
nextcloud-nextcloud-1  |    The user "user1" was created successfully
nextcloud-nextcloud-1  |    The user "nextcloud" was created successfully
nextcloud-nextcloud-1  |    The user "user3" was created successfully
nextcloud-nextcloud-1  |    The user "user5" was created successfully
nextcloud-nextcloud-1  |    The user "jane" was created successfully
nextcloud-nextcloud-1  |    The user "john" was created successfully
nextcloud-nextcloud-1  |    The user "bob" was created successfully
nextcloud-nextcloud-1  |    The user "user4" was created successfully
nextcloud-nextcloud-1  |    The user "user2" was created successfully
```

When I used your project the first time, I didn't understand why I could not log in with the admin account. Because I run `docker-compose up -d proxy nextcloud` and immediately after, I tried to log in. It didn't work because the admin account didn't create ^^'

Author: I created a commit to fix it : https://github.com/juliushaertl/nextcloud-docker-dev/pull/87/commits/e3594fe5c40bd14fc8f007701ce2564567413415

What do you think ?

Of course, you should adapt to the nextcloud release you are using (server, stable23, stable24, and so on.).

If you have not yet generated an app, you can do so from this web page : [https://apps.nextcloud.com/developer/apps/generate](https://apps.nextcloud.com/developer/apps/generate).

## Going further

If you want to go further, you can add new features or customize your development environment by following this documentation : [Configure my environment](docs/configure-vars-env.md).

If you use **XDEBUG** to debug your PHP code. Please, read the [Set Up Xdebug](docs/setup-xdebug.md) documentation.

If you want to set up the Nextcloud's core, please, read the [Manual setup](docs/manual-setup.md) documentation.

If you aren't comfortable with Docker or other tools from the various containers (ldap, mysql, and so on.). You can read the [Useful commands](docs/useful-commands.md) file with some tips.

If you encounter any problems, please, look at this documentation : [Troubleshooting](docs/troubleshooting.md).

### Different feature you can use

These are features where you can use :

> - ☁ Nextcloud

Reviewer 2: Maybe we can add this list and the sections from the readme as a table of contents to the start of the file?

Author: Yes, I agree with you ! And we can add the rest of documentation (`docs/`)

What do you think ?

Author: Okay, I created a table of contents for the READE.md for the moment : https://github.com/juliushaertl/nextcloud-docker-dev/pull/87/commits/9feb2adf57bfdcc9fe0d36ff472c012b5545db13

This project is very **modular** where you can add [features](#different-feature-you-can-use).
So, you can use this project for your development environment only.

> If you don't know how to install Docker, please, read this documentation : https://cloud.nextcloud.com/s/iyNGp8ryWxc7Efa .

Reviewer 2: ```suggestion
If you don't know how to install Docker, please, read this documentation: https://cloud.nextcloud.com/s/iyNGp8ryWxc7Efa
```

Author: Thanks Julius :)

> Be careful : Read the "README.md" first.

Reviewer 2: Might be a bit confusing that it is about the repo readme ;)

Author: Yes, I know, I thought so too ^^

Author: What do you think with this change : https://github.com/juliushaertl/nextcloud-docker-dev/pull/87/commits/61db98eeee42d5afab9ad33d6744e2add679948f ?
